source/agora/test/Byzantine.d

```diff
@@ -0,0 +1,300 @@
+class ByzantineNode (ByzantineReason reason) : TestValidatorNode
+{
+    ByzantineReason reason;
+    public this (Config config, Registry* reg, immutable(Block)[] blocks,
+        ulong txs_to_nominate, shared(time_t)* cur_time)
+    {
+        super(config, reg, blocks, txs_to_nominate, cur_time);
+    }
+
+    protected override TestNominator getNominator (immutable(ConsensusParams) params,
+        Clock clock, NetworkManager network, KeyPair key_pair, Ledger ledger,
+        TaskManager taskman)
+    {
+        return new ByzantineNominator(params, clock, network, key_pair, ledger,
+            taskman, reason);
+    }
+}
+
+static class SpyNominator : TestNominator
+{
+    private shared(size_t)* envelope_count;
+
+    /// Ctor
+    public this (immutable(ConsensusParams) params, Clock clock,
+        NetworkManager network, KeyPair key_pair, Ledger ledger,
+        TaskManager taskman, ulong txs_to_nominate,
+        shared(time_t)* curr_time, shared(size_t)* envelope_count)
+    {
+        this.envelope_count = envelope_count;
+        super(params, clock, network, key_pair, ledger, taskman, txs_to_nominate);
+    }
+
+    public override void receiveEnvelope (scope ref const(SCPEnvelope) envelope) @trusted
+    {
+        atomicOp!("+=")(*this.envelope_count, 1);
+        writeln("SpyNominator.receiveEnvelope");
+    }
+}
+
+static class SpyingValidator : TestValidatorNode
+{
+    public shared(size_t)* envelope_count;
+
+    /// Ctor
+    public this (Config config, Registry* reg, immutable(Block)[] blocks,
+                    ulong txs_to_nominate, shared(time_t)* cur_time,
+                    shared(size_t)* envelope_count)
+    {
+        this.envelope_count = envelope_count;
+        super(config, reg, blocks, txs_to_nominate, cur_time);
+    }
+
+    ///
+    protected override TestNominator getNominator (
+        immutable(ConsensusParams) params, Clock clock,
+        NetworkManager network, KeyPair key_pair, Ledger ledger,
+        TaskManager taskman)
+    {
+        return new SpyNominator(
+            params, clock, network, key_pair, ledger, taskman,
+            this.txs_to_nominate, this.cur_time, this.envelope_count);
+    }
+}
+
+/// create some nodes depending which will not sign or will sign with invalid signature
+static class ByzantineManager (bool addSpyValidator = false,
+    size_t byzantine_not_signing_count = 0,
+    size_t byzantine_bad_signing_count = 0) : TestAPIManager
+{
+    shared size_t envelope_count;
+    ///
+    public this (immutable(Block)[] blocks, TestConf test_conf, time_t genesis_start_time)
+    {
+        super(blocks, test_conf, genesis_start_time);
+    }
+
+    public override void createNewNode (Config conf,
+        string file = __FILE__, int line = __LINE__)
+    {
+        if (this.nodes.length < byzantine_not_signing_count + byzantine_bad_signing_count)
+        {
+            auto time = new shared(time_t)(this.initial_time);
+            assert(conf.node.is_validator);
```

Comment:
I'm not sure why want to check the is_validator  here.
is_validator has been moved to NetworkManager.

Reply:
The `is_validator` boolean has been refactored [here](https://github.com/bpfkorea/agora/blob/v0.x.x/source/agora/common/Config.d#L160-L164):
If you are reading from a `Config`, you can use:
```suggestion
            assert(conf.validator.enabled);
```

Reply:
I missed the latest trend.Sorry
This is it. `conf.validator. enabled`

Reply:
Thanks guys will update